src/wayland/drm_syncobj/mod.rs

```diff
@@ -0,0 +1,415 @@
+        _data: &(),
+        _dh: &DisplayHandle,
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        match request {
+            wp_linux_drm_syncobj_manager_v1::Request::GetSurface { id, surface } => {
+                let already_exists = with_states(&surface, |states| {
+                    states
+                        .data_map
+                        .get::<RefCell<Option<WpLinuxDrmSyncobjSurfaceV1>>>()
+                        .map(|v| v.borrow().is_some())
+                        .unwrap_or(false)
+                });
+                if already_exists {
+                    resource.post_error(
+                        wp_linux_drm_syncobj_manager_v1::Error::SurfaceExists as u32,
+                        "the surface already has a syncobj_surface object associated".to_string(),
+                    );
+                    return;
+                }
+                let syncobj_surface = data_init.init::<_, _>(
+                    id,
+                    DrmSyncobjSurfaceData {
+                        surface: surface.downgrade(),
+                    },
+                );
+                with_states(&surface, |states| {
+                    states
+                        .data_map
+                        .insert_if_missing(|| RefCell::new(Some(syncobj_surface)))
+                });
+                compositor::add_pre_commit_hook::<D, _>(&surface, commit_hook);
+                compositor::add_destruction_hook::<D, _>(&surface, destruction_hook);
+            }
+            wp_linux_drm_syncobj_manager_v1::Request::ImportTimeline { id, fd } => {
+                match DrmTimeline::new(&state.drm_syncobj_state().import_device, fd.as_fd()) {
+                    Ok(timeline) => {
+                        data_init.init::<_, _>(id, DrmSyncobjTimelineData { timeline });
+                    }
+                    Err(err) => {
+                        resource.post_error(
+                            wp_linux_drm_syncobj_manager_v1::Error::InvalidTimeline as u32,
+                            format!("failed to import syncobj timeline: {}", err),
+                        );
+                    }
+                }
+            }
+            wp_linux_drm_syncobj_manager_v1::Request::Destroy => {}
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// Data attached to wp_linux_drm_syncobj_surface_v1 objects
+#[derive(Debug)]
+pub struct DrmSyncobjSurfaceData {
+    surface: Weak<WlSurface>,
+}
+
+impl<D> Dispatch<WpLinuxDrmSyncobjSurfaceV1, DrmSyncobjSurfaceData, D> for DrmSyncobjState {
+    fn request(
+        _state: &mut D,
+        _client: &Client,
+        _resource: &WpLinuxDrmSyncobjSurfaceV1,
+        request: wp_linux_drm_syncobj_surface_v1::Request,
+        data: &DrmSyncobjSurfaceData,
+        _dh: &DisplayHandle,
+        _data_init: &mut DataInit<'_, D>,
+    ) {
+        let Ok(surface) = data.surface.upgrade() else {
+            return;
```

Comment:
Isn't this potentially a `NoSurface` error?

Reply:
Ah true, it does look like the protocol spec states this should return a `no_surface` for `set_acquire_point` or `set_release_point`.